packages/earljs/src/isEqual/getCanonicalType.ts

```diff
@@ -0,0 +1,46 @@
+export type CanonicalType = ReturnType<typeof getCanonicalType>
+
+export function getCanonicalType(value: unknown) {
+  const primitive = getBasicType(value)
+  if (primitive) {
+    return primitive
+  } else if (Array.isArray(value)) {
+    return 'Array'
+  } else if (value instanceof Error) {
+    return 'Error'
+  } else if (value instanceof Date) {
+    return 'Date'
+  } else if (value instanceof RegExp) {
+    return 'RegExp'
+  } else if (value instanceof Map) {
+    return 'Map'
+  } else if (value instanceof Set) {
+    return 'Set'
+  } else if (value instanceof Promise) {
+    return 'Promise'
+  } else if (value instanceof String) {
+    return 'String'
+  } else if (value instanceof Number) {
+    return 'Number'
+  } else if (value instanceof Boolean) {
+    return 'Boolean'
+  } else if (value instanceof WeakMap) {
+    return 'WeakMap'
+  } else if (value instanceof WeakSet) {
+    return 'WeakSet'
+  }
+  return 'Object'
+}
+
+function getBasicType(value: unknown) {
```

Comment:
A few nitpicks.

1. I'd add explicity `return undefined`, because at the first look `const primitive = getBasicType(value); if (primitive) {}` above seems to be always truthy in the if condition.

2. Since this logic is kind of important and would nicely fit the top of the file, wouldn't it be better to inline in `getCanonicalType`?

3. I had fell into a pot with x86 assembly and C++ years ago, and I just noticed `const type = typeof value` in your code.  
  Brace yourself, here comes the story.
  
   Assigning the result `typeof x` expression to a variable and comparing this variable `type === "string"` produces different bytecode in V8 than comparing `typeof x` directly with a constant string `typeof x === "string"`.
  
   Take a look here
  
   ```js
   // indirect.js
   function isString(x) {
     const type = typeof x
     return type === 'string'
   }
   ```
   
   ```sh
   ❯ node --print-bytecode --print-bytecode-filter=isString indirect.js
   [generated bytecode for function: isString (0x00f812e11091 <SharedFunctionInfo isString>)]
   Parameter count 2
   Register count 1
   Frame size 8
      38 S> 000000F812E11A16 @    0 : 25 02             Ldar a0
            000000F812E11A18 @    2 : 52                TypeOf
            000000F812E11A19 @    3 : 26 fb             Star r0
      49 S> 000000F812E11A1B @    5 : 12 00             LdaConstant [0]
      61 E> 000000F812E11A1D @    7 : 68 fb 00          TestEqualStrict r0, [0]
      73 S> 000000F812E11A20 @   10 : aa                Return
   Constant pool (size = 1)
   000000F812E119C9: [FixedArray] in OldSpace
    - map: 0x021b5ad80729 <Map>
    - length: 1
              0: 0x021b5ad840e9 <String[#6]: string>
   Handler Table (size = 0)
   Source Position Table (size = 10)
   0x00f812e11a29 <ByteArray[10]>
   ```
  
   ```js
   // direct.js
   function isString(x) {
     return typeof x === 'string'
   }
   ```
  
   ```
   ❯ node --print-bytecode --print-bytecode-filter=isString direct.js
   [generated bytecode for function: isString (0x01fbc9e11091 <SharedFunctionInfo isString>)]
   Parameter count 2
   Frame size 0
      25 S> 000001FBC9E119FE @    0 : 25 02             Ldar a0
            000001FBC9E11A00 @    2 : 73 01             TestTypeOf #1
      53 S> 000001FBC9E11A02 @    4 : aa                Return
   Constant pool (size = 0)
   Handler Table (size = 0)
   Source Position Table (size = 6)
   0x01fbc9e11a09 <ByteArray[6]>
   ```
  
  So instead of string comparison with `TestEqualStrict`, we get `TestTypeOf`. Pretty lit, right?

  I remember an ESLint rule forbidding comparison with saved typeof string existed, but I can't recall the name :D

Reply:
Glad to have someone dig deep enough to prove something I suspected but didn't know for sure. I took your suggestion and inlined the code.